### ISSUE TITLE
Addition of generic pre build and post build hooks.

### ISSUE THREAD

**Issue by erikd** [haskell/cabal#9892]
**Describe the feature request**
The ability to call shell scripts just before and just after each package is built.

At IOG, our use case for these is for build caching, particularly in CI. Since these pre and post hooks are just shell scripts, there are probably other uses for them.

**Additional context**
At IOG we have a number of very large Haskell projects with deep dependency trees, that can take a long time to build in CI.

The obvious answer to long build times is caching of build products. A previous attempt at this caching [was made](https://hackage.haskell.org/package/cabal-cache), but that solution was not really very satisfactory, because the cache was keyed on `${CPU}-${OS}-${GHC_VERSION)-${hash-of-dependencies}`. The first three are obvious. The problem is `hash-of-dependencies`. If a single high level dependency changes, there will be no cache hit and everything will be built from scratch. As it turns out, this is actually the most common scenario.

A better caching solution is one where the caching is done on individual dependencies rather than on all the dependencies as a huge blob. Caching individual dependencies means that when a high level dependency changes, there is a very high likelihood that all the lower level dependencies will still be found in the cache.

My initial implementation of this package level caching was done as a simple wrapper around `cabal` that used `rsync` to fetch and save the cache over `ssh` to another machine. This proved highly effective and I was able to populate the cache from one machine and use if from another (both machines running Debian Linux).

However, @angerman came up with an even better solution that required adding the ability to run shell scripts before and after the build of each individual package. Using this feature (we have rough patches against cabal `HEAD` and version `3.10.3.0`) we are able to use our own Amazon S3 storage for our cache. We do not propose to make this S3 storage public (obvious potential security problems) but any organization like ours or any individual could use their own S3 storage. I have not tried it yet, but I suspect `ssh` to a different machine could also be used for storage.

The naive patch against `HEAD` (error handling could be improved, maybe the hook names could be made configurable) is:
```
diff --git a/cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs b/cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs
index 065334d5c..570c9b18c 100644
--- a/cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs
+++ b/cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs
@@ -678,7 +678,22 @@ buildAndInstallUnpackedPackage
           runConfigure
         PBBuildPhase{runBuild} -> do
           noticeProgress ProgressBuilding
-          runBuild
+          -- run preBuildHook. If it returns with 0, we assume the build was
+          -- successful. If not, run the build.
+          code <- rawSystemExitCode verbosity (Just srcdir) "preBuildHook" [
+              (unUnitId $ installedUnitId rpkg)
+            , (getSymbolicPath srcdir)
+            , (getSymbolicPath builddir)
+            ] `catchIO` (\_ -> return (ExitFailure 10))
+          when (code /= ExitSuccess) $ do
+            runBuild
+            -- not sure, if we want to care about a failed postBuildHook?
+            void $ rawSystemExitCode verbosity (Just srcdir) "postBuildHook" [
+                (unUnitId $ installedUnitId rpkg)
+              , (getSymbolicPath srcdir)
+              , (getSymbolicPath builddir)
+              ] `catchIO` (\_ -> return (ExitFailure 10))
+
         PBHaddockPhase{runHaddock} -> do
           noticeProgress ProgressHaddock
           runHaddock
```
An example of our `preBuildHook` script (kept in `~/.cabal/iog-hooks`, S3 credentials pulled from `s3-credentials.bash`, the `aws` executable is from the `awscli` package) is as follows:
```
#!/usr/bin/env bash

# shellcheck disable=SC1091,SC2046
. $(dirname "${BASH_SOURCE[0]}")/s3-credentials.bash

CACHE_KEY="$1.tar.gz"

# Check if artifact exists in S3 (with endpoint and credentials)
aws s3 ls s3://"$CACHE_BUCKET"/"$CACHE_KEY" --endpoint-url "$AWS_ENDPOINT" > /dev/null 2>&1

# shellcheck disable=SC2181
if [ $? -eq 0 ]; then
  echo "S3 hit       $1"
  aws s3 cp s3://"$CACHE_BUCKET"/"$CACHE_KEY" - --endpoint-url "$AWS_ENDPOINT" | tar -xz
else
  echo "S3 miss      $1"
  exit 1
fi
```
To use the cache I run `cabal` as:
```
PATH=$PATH:$HOME/.cabal/iog-hooks cabal 
```

**Comment by erikd** [comment]
Example output using the `ssh` storage backend:
```
...
Building     hsc2hs-0.68.10 (exe:hsc2hs)
Building     hspec-core-2.11.7 (lib)
SSH miss     optparse-applicative-0.18.1.0-1068e86b013a3ad1d9c7c30a1b2f110becd16c2a9b2bafbbbbf495dde8e95208
SSH miss     unliftio-0.2.25.0-986d17855f4a377ca1eed1f45d7be8be605e99e81c9ba8c32f42e86c61d9d6bb
SSH miss     hsc2hs-0.68.10-e-hsc2hs-4f2bbae1b93e7038909e0edd772f9551b21bb7c35ac6e6e1ec5eb4bc661f09a2
SSH miss     hspec-core-2.11.7-ad6c16f02379c016ce6c55335b79336fc87a8c13fc2440e30d04b1b849c508c2
SSH upload   recursion-schemes-5.2.2.4-be95a42f55d2ff5478445dafd7fe727dae50fa389f389bc38f05105c5ff7f49e
Installing   recursion-schemes-5.2.2.4 (lib)
Completed    recursion-schemes-5.2.2.4 (lib)
Configuring library for cardano-git-rev-0.2.2.0..
Preprocessing library for cardano-git-rev-0.2.2.0..
...
```

**Comment by fgaz** [comment]
Related: #7394

**Comment by erikd** [comment]
I just did a test using `ghc-9.6` on a machine with 12 cores and 48 of RAM (ie significantly more powerful than expected CI build machines).

With an empty cache:
```bash
cabal clean
rm -rf ~/.cabal/store/ghc-9.6.4/
time PATH=$PATH:$HOME/.cabal/iog-hooks/ cabal build all

real    29m53.885s
user    105m12.177s
sys     11m5.029s
```
With a warm cache (ie immediately after the above):
```bash
cabal clean
rm -rf ~/.cabal/store/ghc-9.6.4/
time PATH=$PATH:$HOME/.cabal/iog-hooks/ cabal build all


real    13m20.817s
user    37m24.021s
sys     6m58.173s
```

**Comment by mpickering** [comment]
Is there an advantage to doing this using a hook rather than populating the caches based on `plan.json` before starting to do any building?

It seems that using an approach like [`cabal-cache`](https://github.com/haskell-works/cabal-cache), using `plan.json` would be easier and wouldn't require modifying cabal-install.

**Comment by newhoggy** [comment]
If I read this proposal correctly, it will fix this issue, correct?

https://github.com/haskell-works/cabal-cache/issues/166

**Comment by newhoggy** [comment]
There are some other issue encountered, such as packages having absolute paths built into them which mean them non-transferrable to other machines where the build location, store location or some other environment differs.

**Comment by newhoggy** [comment]
> Is there an advantage to doing this using a hook rather than populating the caches based on plan.json before starting to do any building?

Brainstorming a bit:

I can see this potentially evolving to the point where in CI or even on development machines we don't build any of the dependencies because we can point to a caching build server.  If we reach this point the advantage is that if we have a large projects, components that depend on already cached dependencies can start building earlier whilst the remote build server is busy populating the cache for yet unbuilt dependencies.

It could be a paid for cloud service as well, maybe something the Haskell Foundation can run?

The advantage of this would be minimal idle time by build client.

**Comment by angerman** [comment]
There are a few advantages to hooks that you do not get from reconstructing the store outside of cabal. As @erikd showed the patch is currently also ~20lines of extra code in cabals codebase and it permits all kinds of use for the hooks. You could run some time profiling with them, or anything else you'd like to do around builds.

- By being at the location where it is, and effectively substituting GHC's build, you can omit _Paths' modules from the reconstruction, and have GHC build paths modules for you.
- It's not trying to reconstruct the store, and leaves all the store logic (layout, ...) up to cabal. Inlcuding .conf generation, copying and installing. (Most of the logic of cabal is retained in cabal, and the hooks do not have to concern themselves with any of that.

ABI is a major issue. Luckily for us we are getting close to making it irrelevant for _us_.

**Comment by angerman** [comment]
One last note: these are _generic_ hooks. Caching may be _one_ use for them. But hooks into cabal have been requested (and discussed outside of issue trackers) multiple times.

**Comment by newhoggy** [comment]
Can you elaborate on the `_Paths` modules?  Are they still going to be included in the packages but ghc patches after the fact?  Or will they be gone from the packages?

**Comment by angerman** [comment]
@newhoggy caching is clearly only _one_ instance of what _could_ be done with these hooks. (And we likely want cabal to have more hooks!).

> Can you elaborate on the _Paths modules? Are they still going to be included in the packages but ghc patches after the fact? Or will they be gone from the packages?

You can just _not_ backup any _Paths modules in the tarball you build `--exclude=*_Paths.*` or similar.

When you restore the cached objects into the build folder ghc will just build the ones that are missing. After that cabal will continue with copy/reg/install.

> I wanted to understand if pre/post build hooks are the only ones we need. For example do we need a no build hook because the package was already built?

For this PoC, the pre/post hooks seem to be enough. However, you'd likely want to have many more hooks in cabal to hook different phases for experimentation, and extensions.

**Comment by hasufell** [comment]
> Related: #7394
> 
> I think this would be a welcome feature, especially if it's written with the above issue in mind so it can be extended to other actions.

Exactly. This has been requested/discussed since the dawn of time.

Hooks are as old as unix and the other major design pattern after "pipes". I'm not sure it even warrants a discussion... and we surely shouldn't be discussing a specific use case. There are infinite.

The question is more about:

* where exactly do we want to insert hooks
* what information should be exposed to the hooks

I think last time it was discussed, it wasn't really clear where to insert hooks, because cabal had no clean architecture that resembles strictly the configure/build/install phases that traditional package managers have.

So this would be the only concern I see: which parts are stable enough architecturally that they won't break hooks in the future. Or do we need a refactor first?

**Comment by michaelpj** [comment]
One obvious concern with this is portability, right? At the moment most of the stuff that cabal does is limited to a known set of things that cabal handles and can be tested. Users can't currently inject random non-portable stuff, but these kinds of hooks would make it very easy to do that.

(Of course you _can_ do this today with `build-type: Custom` or TH or something, but it's perhaps a bit trickier?)

**Comment by hasufell** [comment]
> One obvious concern with this is portability, right? 

I'm not sure I don't understand this statement.

User hooks are not in scope to be tested by cabal. It's irrelevant what users do with them. If a user manages to write a package that only works with some user hooks supplied, then that's simply a broken package.

**Comment by hasufell** [comment]
>  meant that it's easier for users to accidentally make packages that are non-portable

I don't see how. Can you give an example? The shell hook doesn't change any internal data structures of cabal. The worst it could do is generate files that are needed for compilation, in which case **it is a broken package**.

I don't see why users would think shell hooks are the right place to do this.

**Comment by geekosaur** [comment]
Because users will find a way to abuse anything you give them? &lt;/sysadmin&gt;

**Comment by angerman** [comment]
> I meant that it's easier for users to accidentally make packages that are non-portable, since now they have to be sure those shell scripts will run in all situations where a downstream user is compiling their package.

If someone is relying on some random hooks to exist at some downstream user for their package, they are arguably doing it wrong. You do not ship hooks with your package.

Hooks are end user customizations, nothing that package authors should consider. This is _not_ BuildHooks.

As an emacs package author you also do not rely on, or test all possible combinations of, hooks an end user might put into their config.

Adding hooks to _your_ cabal install means you (the end user) get a modified cabal. That's on you. Always.

**Comment by erikd** [comment]
> Is there an advantage to doing this using a hook rather than populating the caches based on `plan.json` before starting to do any building?

Yes, the hook version is twice as fast as the populating cache based on `plan.json` and using `rsync` to move files to and from a remove machine.

Pretty sure this is because the hook version interleaves S3 caching operations while the `rsync` solution does the rsync first (for cache restore) and then build.

**Comment by erikd** [comment]
> Does this issue propose to fix this?

I believe so, yes. The cache uses `<package-name>-<package-version>-<config-hash>` as the cache key. For a given `<package-name>-<package-version>` the `<config-hash>` will be different for any single one of different sets of cabal flags, different CPUs, different OSes etc.

**Comment by gbaz** [comment]
Not suggesting this per-se, but would it be possible to "fake" this particular set of hooks by wrapping ghc itself?

**Comment by angerman** [comment]
> Not suggesting this per-se, but would it be possible to "fake" this particular set of hooks by wrapping ghc itself?

Yes. Conceptually you could. Though it's harder to say _why_ you were called if you wrap GHC. You could try to use some heuristics to determine the caller, hooks are more explicit in that scenario.

**Comment by erikd** [comment]
> Not suggesting this per-se, but would it be possible to "fake" this particular set of hooks by wrapping ghc itself?

The problem with wrapping GHC is as follows. Consider two projects, one where you do want to use hooks and one where you don't. How does a wrapper GHC tell the difference? With the hooks version the difference is inherent.

**Comment by geekosaur** [comment]
I find myself wondering how that would interact with other things that wrap ghc (such as HLS) as well.

**Comment by gbaz** [comment]
> The problem with wrapping GHC is as follows. Consider two projects, one where you do want to use hooks and one where you don't. How does a wrapped GHC tell the difference? With the hooks version the difference is inherent.

Well you pass the ghc per-project with the --with-ghc flag, which can go into a project file, just like a build hook can, I imagine.

I'm open to the hooks concept in general, but it seems like in this use case the "real" ask is "a ghc with high-grade distributed caching as part of the build process" so it feels more natural to me to 'fake' that directly.

**Comment by erikd** [comment]
> Well you pass the ghc per-project with the --with-ghc flag, which can go into a project file, just like a build hook can, I imagine.

The `--with-ghc` option in the project file would not work for us because all out projects (about 8 of them) build with multiple versions of `ghc` from `8.10.7` through to `9.8.1`. We even build in CI with at least two versions of GHC.

**Comment by angerman** [comment]
The bigger issue with `--with-ghc` is that it just globally gives the compiler to cabal. If you end up being called from cabal, you still don't know why. It could be for building, or for something else (cabal also likes to interrogate it's tools).

**Comment by andreabedini** [comment]
IMHO the best approach would be to better separate planning and building phases, so that you could swap the build project with your own (or part of it, or add hooks, or whatever).
That said, this is not a new idea and I think recent refactors made it easier to implement.

I think there are some considerations to make [1] but we should not bikeshed it to death.
@erikd I understand you have poc already, could you open a PR (even draft) so we can discuss the details? 

e.g. I assume this is _build_ as in _./Setup.hs build_, compiling Setup.hs is a separate thing.

**Comment by erikd** [comment]
@andreabedini The complete patch (against `HEAD`) is in the original post. I will raise it as a PR.

**Comment by erikd** [comment]
PR is https://github.com/haskell/cabal/pull/9899 . Only a draft, but still needs documentation and changelog entry etc.

**Comment by hasufell** [comment]
My main interest in hooks is interfacing with GHCup.

Which phase of cabal is early enough that:

- we already know what GHC to use for the build
- we haven't attempted solving or any configuration steps yet

This will allow users to emulate stack behavior, where running `cabal build` may download the right GHC version and possibly HLS. It would also potentially allow to remove lots of code from the vscode-haskell extension, which hacks around the lack of this in other ways.

**Comment by erikd** [comment]
> Which phase of cabal is early enough that:

These hooks I am proposing are not a solution to the problem you want to solve.

**Comment by erikd** [comment]
I have been playing around with hooks located in `~/.cabal/hooks/` for  couple of weeks and quite frankly they are a pain in the neck.

Hooks should be project specific and not global. For instance if I have some work projects and some personal projects, I may want to have different hooks for each, or have hooks for one and not for the other.

To me it makes more sense for `cabal` to look in the `cabalHooks` directory (with this directory being in the same directory as the `cabal.project` file) although @angerman would prefer the hook directory to be named `.cabal/hooks` like the way `git` hooks are located in `.git/hooks/`.

**Comment by erikd** [comment]
After using this for a couple of days, this project local version is **VASTLY** better than the user global version.

I have updated the PR: https://github.com/haskell/cabal/pull/9899

**Comment by ulysses4ever** [comment]
@erikd given your last message on the PR #9899, do you think this issue should be closed as well?

**Comment by angerman** [comment]
I've had some discussion with @erikd around security. I'll let him know he should leave a comment here.

**Comment by ulysses4ever** [comment]
Thanks @angerman!

**Comment by erikd** [comment]
@angerman and I have had some discussion about how to make hooks secure. We think it can be made secure enough for me to think its acceptable.

My current idea is that we add a file `~/.cabal/hook-security` to the user's home directory. This file look something like:
```
trust-always:
  /home/user/work/project/hooks/first.sh

trust-sha256:
  /home/user/work/project/hooks/second.sh <sha256>
```
  
This `hook-security` file should be in the users home directory and not part of a git checkout (and hence not updated with a `git checkout` oepration).

**Comment by angerman** [comment]
Just one UX note on this, which got lost:

> Hook files not listed in this file would not be trusted and not run.

Cabal would _inform_ the user about the existing hook in the repo / or changed hash of the hook. And ask if the user wants to trust it interactively if we are in an interactive invocation. (y/N).

> There is an untrusted hook: path/to/hook.sh. Do you want to trust it (y)es, (N)o, or (s)how? (y/N/s)

show would just "cat" the hook to stdout. And ask the question again.

**Comment by erikd** [comment]
> > There is an untrusted hook: path/to/hook.sh. Do you want to trust it (y)es, (N)o, or (s)how? (y/N/s)
> 
> show would just "cat" the hook to stdout. And ask the question again.

`cabal` should probably avoid being interactive for cases like when it is being run in CI.
